clusterloader2/pkg/measurement/common/kube_state_metrics_measurement.go

```diff
@@ -0,0 +1,221 @@
+	namespace   string
+	serviceName string
+	metricsPort int
+	selfPort    int
+	latency     []ksmLatencyMetrics
+	wg          sync.WaitGroup
+}
+
+type ksmLatencyMetrics struct {
+	latencyMetrics *measurementutil.Histogram
+}
+
+func init() {
+	if err := measurement.Register(ksmLatencyName, CreateKSMLatencyMeasurement); err != nil {
+		klog.Fatalf("Cannot register %s: %v", ksmLatencyName, err)
+	}
+}
+
+// CreateKSMLatencyMeasurement creates a new Kube State
+// Metrics Measurement.
+func CreateKSMLatencyMeasurement() measurement.Measurement {
+	return &ksmLatencyMeasurement{
+		namespace:   "kube-state-metrics-perf-test",
+		serviceName: "kube-state-metrics",
+		selfPort:    8081,
+		metricsPort: 8080,
+	}
+}
+
+// Execute supports two actions:
+// - start -  query on /metrics and gathers initial latency data.
+// - gather - gets the latency data again and creates a latency summary.
+func (m *ksmLatencyMeasurement) Execute(config *measurement.Config) ([]measurement.Summary, error) {
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return nil, err
+	}
+	probeDuration := 10 * time.Second
+	client := config.ClusterFramework.GetClientSets().GetClient()
+
+	switch action {
+	case "start":
+		if m.isRunning {
+			klog.V(2).Infof("%s: measurement already running", m)
+			return nil, nil
+		}
+		m.stopCh = make(chan struct{})
+		err = m.start(client, probeDuration)
+		if err != nil {
+			return nil, err
+		}
+		return nil, nil
+	case "gather":
+		return m.getKSMLatency(client, probeDuration)
+	default:
+		return nil, fmt.Errorf("unknown action %v", action)
+	}
+}
+
+// getKSMLatency gathers the latency and creates the summary based on the Quantile from the sub histograms.
+func (m *ksmLatencyMeasurement) getKSMLatency(client clientset.Interface, probeDuration time.Duration) ([]measurement.Summary, error) {
+	if !m.isRunning {
+		return nil, nil
+	}
+	close(m.stopCh)
+	m.wg.Wait()
+	// We want to substract the latest histogram from the first one we collect.
+	finalLatency := HistogramSub(m.latency[len(m.latency)-1].latencyMetrics, m.latency[0].latencyMetrics)
+	// Pretty Print the report.
+	result := &measurementutil.LatencyMetric{}
+	err := SetQuantileFromHistogram(result, finalLatency)
+	if err != nil {
+		return nil, err
+	}
+	content, err := util.PrettyPrintJSON(result)
+	if err != nil {
+		return nil, err
+	}
+	// Create Summary.
+	metricsSummary := measurement.CreateSummary(schedulerLatencyMetricName, "json", content)
+	return []measurement.Summary{metricsSummary}, nil
+}
+
+func (m *ksmLatencyMeasurement) start(client clientset.Interface, probeDuration time.Duration) error {
+	m.isRunning = true
+	m.wg.Add(1)
+	var err error
+	var latHist *measurementutil.Histogram
+	go func() {
+		defer m.wg.Done()
+		for {
+			select {
+			case <-m.stopCh:
+				return
+			case <-time.After(probeDuration):
+				latHist, err = m.retrieveKSMLatencyMetrics(client, probeDuration)
+				m.latency = append(m.latency, ksmLatencyMetrics{
```

Comment:
Why we're storing every single call, given that we seem to be computing the result from the first and last call only?

In other words, based on the getKSMLatency() method, I'm wondering, why the pattern isn't:
- start: "get KSM metrics"
- gather: "get KSM metric, substract those two results"
- ?

Reply:
Currently, we are storing all calls yes, this was mainly as in the future I also want to add a few other things which we need storing for. So I might have prematurely optimized for that in mind. But yes just getting two calls, first and last is enough, can change and in the future when I need to can revert to this?

Reply:
Yeah - we can always extend and start gathering more information if needed.
I wouldn't over-engineer for things that "we may potentially need in the future".